com.microsoft.mrtk.uxcore/Keyboard/KeyboardPreview.cs

```diff
@@ -0,0 +1,411 @@
+            {
+                solverHandler.UpdateSolvers = !solverHandler.UpdateSolvers;
+
+                if (solverHandler.UpdateSolvers)
+                {
+                    ApplyShellSolverParameters();
+                }
+            }
+#endif
+        }
+
+        #region MonoBehaviour Implementation
+
+        private void OnEnable()
+        {
+            StartCoroutine(BlinkCaret());
+        }
+
+        private void Start()
+        {
+            ApplyShellSolverParameters();
+        }
+
+        private void LateUpdate()
+        {
+            if (layoutInvalidated)
+            {
+                layoutInvalidated = false;
+                UpdateLayout();
+            }
+
+        }
+
+        #endregion MonoBehaviour Implementation
+
+        /// <summary>
+        /// Force the <see cref="PreviewText"/> label and <see cref="PreviewCaret"/> transform to update their positions on the next update.
+        /// </summary>
+        public void InvalidateLayout()
+        {
+            layoutInvalidated = true;
+        }
+
+
+        private void UpdateLayout()
+        {
+            if (previewCaret == null || PreviewText == null)
+            {
+                return;
+            }
+
+            var textInfo = PreviewText.GetTextInfo(Text);
+            if (textInfo.lineCount == 0 || textInfo.characterCount == 0)
+            {
+                ResetCaret();
+                ResetView();
+                return;
+            }
+
+            // create a fake previous character if at the start
+            TMP_CharacterInfo prevChar;
+            if (CaretIndex == 0)
+            {
+                prevChar = default;
+                prevChar.index = -1;
+            }
+            else
+            {
+                prevChar = textInfo.characterInfo[CaretIndex - 1];
+            }
+
+            // create a fake next character if at the end
+            TMP_CharacterInfo nextChar;
+            if (CaretIndex == textInfo.characterCount)
+            {
+                nextChar = prevChar;
+                nextChar.character = '\0';
+                nextChar.index++;
+                nextChar.origin = nextChar.xAdvance;
+                if (PreviewText.isRightToLeftText)
+                {
+                    nextChar.topRight.x = nextChar.topLeft.x;
+                    nextChar.bottomRight.x = nextChar.bottomLeft.x;
+                }
+                else
+                {
+                    nextChar.topLeft.x = nextChar.topRight.x;
+                    nextChar.bottomLeft.x = nextChar.bottomRight.x;
+                }
+            }
+            else
+            {
+                nextChar = textInfo.characterInfo[CaretIndex];
+            }
+
+            if (prevChar.lineNumber > textInfo.lineCount)
+            {
+                ResetCaret();
+                ResetView();
+                return;
+            }
+
+            var lineInfo = textInfo.lineInfo[prevChar.lineNumber];
```

Comment:
Per frame allocation.  Can be a field var?

Reply:
no allocation here